compiler/src/dotty/tools/dotc/transform/LazyVals.scala

```diff
@@ -274,171 +288,210 @@ class LazyVals extends MiniPhase with IdentityDenotTransformer {
+   * @param runtimeModule the runtime LazyVals module path
+
+   */
+  def mkThreadSafeDef(memberDef: ValOrDefDef,
                       claz: ClassSymbol,
-                      ord: Int,
                       target: Symbol,
-                      rhs: Tree,
-                      tp: Type,
                       offset: Tree,
-                      getFlag: Tree,
-                      stateMask: Tree,
-                      casFlag: Tree,
-                      setFlagState: Tree,
-                      waitOnLock: Tree)(using Context): DefDef = {
-    val initState = Literal(Constant(0))
-    val computeState = Literal(Constant(1))
-    val computedState = Literal(Constant(3))
-
-    val thiz = This(claz)
-    val fieldId = Literal(Constant(ord))
-
-    val flagSymbol = newSymbol(methodSymbol, lazyNme.flag, Synthetic, defn.LongType)
-    val flagDef = ValDef(flagSymbol, getFlag.appliedTo(thiz, offset))
-    val flagRef = ref(flagSymbol)
-
-    val stateSymbol = newSymbol(methodSymbol, lazyNme.state, Synthetic, defn.LongType)
-    val stateDef = ValDef(stateSymbol, stateMask.appliedTo(ref(flagSymbol), Literal(Constant(ord))))
-    val stateRef = ref(stateSymbol)
-
-    val compute = {
-      val resultSymbol = newSymbol(methodSymbol, lazyNme.result, Synthetic, tp)
-      val resultRef = ref(resultSymbol)
-      val stats = (
-        ValDef(resultSymbol, rhs) ::
-        ref(target).becomes(resultRef) ::
-        (nullOut(nullableFor(methodSymbol)) :+
-        setFlagState.appliedTo(thiz, offset, computedState, fieldId))
-      )
-      Block(stats, Return(resultRef, methodSymbol))
-    }
-
-    val retryCase = {
-      val caseSymbol = newSymbol(methodSymbol, nme.DEFAULT_EXCEPTION_NAME, Synthetic | Case, defn.ThrowableType)
-      val triggerRetry = setFlagState.appliedTo(thiz, offset, initState, fieldId)
-      CaseDef(
-        Bind(caseSymbol, ref(caseSymbol)),
-        EmptyTree,
-        Block(List(triggerRetry), Throw(ref(caseSymbol)))
+                      thiz: Tree,
+                      helperModule: TermSymbol,
+                      runtimeModule: String)(using Context): DefDef = {
+    val tp = memberDef.tpe.widen.resultType.widen
+    val waiting = ref(requiredClass(s"$runtimeModule.${lazyNme.RLazyVals.waiting}"))
+    val evaluating = Select(ref(helperModule), lazyNme.RLazyVals.evaluating)
+    val nullValued = Select(ref(helperModule), lazyNme.RLazyVals.nullValued)
+    val objCasFlag = Select(ref(helperModule), lazyNme.RLazyVals.objCas)
+    val methodSymbol = memberDef.symbol.asTerm
+    val rhs = memberDef.rhs
+    val discardSymb = newSymbol(methodSymbol, lazyNme.discard, Method | Synthetic, MethodType(Nil)(_ => Nil, _ => defn.UnitType))
+    val discardDef = DefDef(discardSymb, initBlock(
+      objCasFlag.appliedTo(thiz, offset, evaluating, Select(New(waiting), StdNames.nme.CONSTRUCTOR).ensureApplied)
+        :: Return(unitLiteral, discardSymb) :: Nil))
```

Comment:
What is this `discard` thing? It doesn't seem to exist in the old implementation, and none of the comments mention it.

Reply:
It is basically:
```scala
if current.isInstanceOf[Evaluating] then
  CAS(current, Evaluating, new Waiting)
```
But it does not make sense to make separate method for that. I'm looking if the original author maybe did it for a reason I don't see, but it seems that I can just inline it.